crates/neon/src/types_impl/buffer/types.rs

```diff
@@ -0,0 +1,758 @@
+/// [ArrayBuffer]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer
+/// [Buffer]: https://nodejs.org/api/buffer.html
+#[derive(Debug)]
+#[repr(transparent)]
+pub struct JsTypedArray<T: private::Binary> {
+    local: raw::Local,
+    _type: PhantomData<T>,
+}
+
+impl<T: private::Binary> private::Sealed for JsTypedArray<T> {}
+
+unsafe impl<T: private::Binary> TransparentNoCopyWrapper for JsTypedArray<T> {
+    type Inner = raw::Local;
+
+    fn into_inner(self) -> Self::Inner {
+        self.local
+    }
+}
+
+impl<T: private::Binary> Managed for JsTypedArray<T> {
+    fn to_raw(&self) -> raw::Local {
+        self.local
+    }
+
+    fn from_raw(_env: Env, local: raw::Local) -> Self {
+        Self {
+            local,
+            _type: PhantomData,
+        }
+    }
+}
+
+impl<T: private::Binary> TypedArray for JsTypedArray<T> {
+    type Item = T;
+
+    fn as_slice<'cx, 'a, C>(&self, cx: &'a C) -> &'a [Self::Item]
+    where
+        C: Context<'cx>,
+    {
+        unsafe {
+            let env = cx.env().to_raw();
+            let value = self.to_raw();
+            let info = sys::typedarray::info(env, value);
+
+            slice::from_raw_parts(info.data.cast(), info.length)
+        }
+    }
+
+    fn as_mut_slice<'cx, 'a, C>(&mut self, cx: &'a mut C) -> &'a mut [Self::Item]
+    where
+        C: Context<'cx>,
+    {
+        unsafe {
+            let env = cx.env().to_raw();
+            let value = self.to_raw();
+            let info = sys::typedarray::info(env, value);
+
+            slice::from_raw_parts_mut(info.data.cast(), info.length)
+        }
+    }
+
+    fn try_borrow<'cx, 'b, C>(
+        &self,
+        lock: &'b Lock<'b, C>,
+    ) -> Result<Ref<'b, Self::Item>, BorrowError>
+    where
+        C: Context<'cx>,
+    {
+        unsafe {
+            let env = lock.cx.env().to_raw();
+            let value = self.to_raw();
+            let info = sys::typedarray::info(env, value);
+
+            // The borrowed data must be guarded by `Ledger` before returning
+            Ledger::try_borrow(
+                &lock.ledger,
+                slice::from_raw_parts(info.data.cast(), info.length),
+            )
+        }
+    }
+
+    fn try_borrow_mut<'cx, 'a, C>(
+        &mut self,
+        lock: &'a Lock<'a, C>,
+    ) -> Result<RefMut<'a, Self::Item>, BorrowError>
+    where
+        C: Context<'cx>,
+    {
+        unsafe {
+            let env = lock.cx.env().to_raw();
+            let value = self.to_raw();
+            let info = sys::typedarray::info(env, value);
+
+            // The borrowed data must be guarded by `Ledger` before returning
+            Ledger::try_borrow_mut(
+                &lock.ledger,
+                slice::from_raw_parts_mut(info.data.cast(), info.length),
+            )
+        }
+    }
+}
+
+impl<T: private::Binary> JsTypedArray<T> {
+    pub fn from_array_buffer<'cx, 'a, C>(
```

Comment:
Nit, `'a` can be elided.

Reply:
It might be nice to have one more helper that assumes the entire buffer should be in the typed array. It might be good for that version to have the name `from_array_buffer` and maybe use some other name for this, e.g., `from_array_buffer_slice`.

```rust
impl<T: private::Binary> JsTypedArray<T> {
    pub fn from_array_buffer<'cx, C>(cx: &mut C, buffer: Handle<JsArrayBuffer>) -> JsResult<'cx, Self>
    where
        C: Context<'cx>,
    {
        let len = buffer.as_slice(cx).len();

        Self::from_array_buffer_slice(cx, buffer, 0, len)
    }
    
    /* ... */
}
```

Reply:
Would it make sense to use Rust's range syntax for this, and have it be a single method?

```rust
JsUint32Array::from_array_buffer(cx, buffer, ..) // full buffer
JsUint32Array::from_array_buffer(cx, buffer, 0 .. len) // buffer from half-open range [0, len)
JsUint32Array::from_array_buffer(cx, buffer, i ..) // buffer from index i to end
```

Reply:
Also I wonder if the `array_` is just unnecessary in the name:

```rust
JsUint32Array::from_buffer(cx, buffer, ..)       // full buffer
JsUint32Array::from_buffer(cx, buffer, 0 .. len) // buffer from half-open range [0, len)
JsUint32Array::from_buffer(cx, buffer, i ..)     // buffer from index i to end
```

On the one hand, Node does have a legacy `Buffer` type, but:
- it's used less often than it used to be since JS typed arrays are standard and more fully-featured
- the standard JS API already has the `.buffer` property which returns an `ArrayBuffer`
- `.from_buffer` reads nicer IMO

Reply:
I like dropping `array`. I'm not sure about range syntax. It looks awkward, but it does make expressing the two optional arguments nicely.

Something to keep in mind is that a range makes it a little complicated because there are a bunch of different range types (Range, FullRange, etc.).

It would be nice to put the range on the buffer as an index, but since that returns a reference, I'm not sure how that would work.

Reply:
Actually, looks like there's a trait that covers all of them.

https://doc.rust-lang.org/std/ops/trait.RangeBounds.html

Reply:
Right, I was thinking maybe we make `from_buffer` take a generic `impl RangeBounds` and `.buffer()` could return `(Handle<ArrayBuffer>, Range)`.

Reply:
Alternatively we could have `from_buffer()` and `buffer()` just take and return the ArrayBuffer, and do what I said above but call it `from_buffer_range()` and `buffer_range()`.

Reply:
There’s some precedent in std like `String::replace_range()`: https://doc.rust-lang.org/alloc/string/struct.String.html#method.replace_range

Reply:
Plot twist! Both the standard JS APIs and Node-API use (_byte offset_ x _number of typed elements_), so a single range has an impedance mismatch. Hmmmm…

Reply:
@kjvalencik I think I'm coming back around to your original suggestion. I'm not sure about the word "slice" since it sort of suggests the input should be a Rust slice datatype. I thought of "range" (which led me to this exploration of range syntax), but that sort of suggests the input should be a Rust range datatype.

`from_buffer_region`?

Or do you think the benefit of the "slice" analogy outweighs the inconsistency/confusion from not actually taking a slice?

Reply:
I went with `from_buffer_region`. I think it reads pretty nicely.